src/sentry/api/serializers/snuba.py

```diff
@@ -0,0 +1,282 @@
+                'dateCreated': eu.date_added,
+                'label': eu.get_label(),
+                'name': eu.get_display_name(),
+                'geo': geo_by_addr(eu.ip_address),
+            },
+        }
+    return rv
+
+
+def encoder_eventuser(value):
+    tag_value, project_id = value
+    return '%d:%s' % (project_id, tag_value)
+
+
+def serialize_projects(organization, item_list, user):
+    return {
+        id: {
+            'id': id,
+            'slug': slug,
+        }
+        for id, slug in Project.objects.filter(
+            id__in=item_list,
+            organization=organization,
+            status=ProjectStatus.VISIBLE,
+        ).values_list('id', 'slug')
+    }
+
+
+def serialize_noop(organization, item_list, user, lookup):
+    return {
+        i: {
+            HEALTH_ID_KEY: make_health_id(lookup, [i[0]]),
+            'value': i[0],
+        }
+        for i in item_list
+    }
+
+
+def encoder_noop(row):
+    if not row:
+        return None
+    return row[0]
+
+
+def value_from_row(row, tagkey):
+    return tuple(row[k] for k in tagkey)
+
+
+def zerofill(data, start, end, rollup):
+    rv = []
+    start = ((int(to_timestamp(start)) / rollup) * rollup) - rollup
+    end = ((int(to_timestamp(end)) / rollup) * rollup) + rollup
+    i = 0
+    for key in xrange(start, end, rollup):
+        try:
+            if data[i][0] == key:
+                rv.append(data[i])
+                i += 1
+                continue
+        except IndexError:
+            pass
+
+        rv.append((key, []))
+    return rv
+
+
+class SnubaLookup(object):
+    __slots__ = 'name', 'tagkey', 'columns', 'selected_columns', 'conditions', 'serializer', 'encoder', 'filter_key'
+    __registry = {}
+
+    def __init__(self, name, tagkey=None, extra=None, selected_columns=None,
+                 conditions=None, serializer=serialize_noop, encoder=encoder_noop,
+                 filter_key=None):
+        cls = type(self)
+        assert name not in cls.__registry
+        self.name = name
+        self.tagkey = tagkey or name
+        self.columns = [self.tagkey] + list(extra or [])
+        self.serializer = partial(serializer, lookup=self)
+        self.encoder = encoder
+        self.conditions = conditions or [[self.tagkey, 'IS NOT NULL', None]]
+        self.selected_columns = selected_columns or []
+        self.filter_key = filter_key or self.tagkey
+        cls.__registry[name] = self
+
+    @classmethod
+    def get(cls, name):
+        return cls.__registry[name]
+
+
+SnubaLookup(
+    'user',
+    'tags[sentry:user]',
+    ['project_id'],
+    serializer=serialize_eventusers,
+    encoder=encoder_eventuser,
+    filter_key=(
+        'concat', (('toString', ('project_id',)), "':'", 'tags[sentry:user]')
+    ),
+)
+SnubaLookup('release', 'tags[sentry:release]', serializer=serialize_releases)
+SnubaLookup('os.name', 'tags[os.name]')
+SnubaLookup('browser.name', 'tags[browser.name]', conditions=[])
```

Comment:
can we add `tags[transaction]`?

Reply:
Yes, but lemme follow up with these. I want to get this merged as-is, then adding on will be simple.

Reply:
Added in e447d0b492a8f6a62891271c63a7f77d024cc3a6.